### PR TITLE
Extend harmonic centrality to include subset

### PR DESCRIPTION
**Pull request by ChristopherReinartz**
Integrated all changes in harmonic_centrality_subset into the original harmonic.py module and moved all tests for the harmonic_subset calculations into test_harmonic_centrality.py.

### CODE REVIEW

**Review by rossbar — COMMENTED** [review]
Sorry if this is a silly question, I'm probably just missing something, but could something similar be achieved with the existing `harmonic_centrality` and passing in a subgraph instead of adding the `nbunch_v` kwarg? I.e. are the following equivalent: `harmonic_centrality(G, nbunch, nbunch_v)` and `harmonic_centrality(nx.subgraph(G, nbunch_v), nbunch)`?

**Comment by dschult** [comment]
I think Ross' idea might be correct.  If so, the tests here should stay, most of the code changes can be removed and the doc_string should add an example where a subgraph is created and used. 

Thanks for this idea!

**Comment by ChristopherReinartz** [comment]
@rossbar It's a good question and I might be misunderstanding what exactly subgraph does, but I think that `harmonic_centrality(G, nbunch, nbunch_v)` and `harmonic_centrality(nx.subgraph(G, nbunch_v), nbunch)` actually produce different results. 
Imagine an edge case where `nbunch` and `nbunch_v` are disjoint sets of nodes. In that case,  `harmonic_centrality(nx.subgraph(G, nbunch_v), nbunch)`, should'nt work because nodes in `nbunch` are not part of `nx.subgraph(G, nbunch_v)`.
`harmonic_centrality(G, nbunch, nbunch_v)` should still produce a result if there exist paths between nodes in `nbunch` and `nbunch_v` in the original graph, because the path length is calculated using the original graph in `harmonic_centrality(G, nbunch, nbunch_v)`.

**Comment by dschult** [comment]
Perhaps we need new descriptions of what this subset handling does then. And better variable names than `nbunch` and `nbunch_v`.  Currently the docs say:
```If nbunch_v is given as an arguemnt, the harmonic centrality for u in
    nbunch is calculated considering only nodes v in nbunch_v.```

That does not align with taking paths through nodes not in `nbunch_v`.
Perhaps `nbunch_v` should be called `sources` or `source_nodes` with 
documentation like:  
```If `sources` is a container of nodes, the harmonic centrality is computed by only summing over reciprocal distances from these source nodes. Otherwise the sum is over all nodes.```

Do you have a citation for this measure?  It doesn't seem to measure centrality so much as distance from a set of nodes. Perhaps some more explanation, description or citation is needed to make this concept clear.

**Comment by ChristopherReinartz** [comment]
@dschult Renaming nbunch_v to sources and nbunch to targets seems like a good idea, if it is okay to break backwards compatibility here. I can change the documentation based on the changes you proposed in your last comment. 
Unfortunately, I don't have a citation or other reference for this. It is just something I need to use frequently lately.

**Comment by dschult** [comment]
Maybe this should be a separate function then -- that is, maybe called `partial_harmonic_centrality` or something better.

What do you need to use it for? Maybe there is a literature of papers that have talked about it?

**Comment by ChristopherReinartz** [comment]
Sorry, I just realized that I opened a second pull request for this topic by accident (originally pull request #4072). 
I originally had it as a separate function, but added it to the existing harmonic.py based on our discussion in #4072. 
I think it's probably easier to find if it is an option in the existing function rather than a separate one, but I can change it back if you like. 

I use it when I have nodes with different attributes and I want to check how 'central' nodes with one attribute are when only considering nodes that have a certain, different attribute.

**Comment by dschult** [comment]
Thanks for that reminder. I recall the discussion now. Somehow I thought it would be less of a change in the routine. It's not even clear to me that calling it with `nbunch_v=G.nodes` will give the same results as `nbunch_v=None` the way the code is written.  Certainly calling the function to find the shortest path between each pair (u, v) is very likely to be slower than getting all_pairs and eliminating those not in nbunch_v.

Could this change be done (and what's the impact on speed) with
```python
if nbunch_v is None:
    nbunch_v = G.nodes
else:
    nbunch_v = set(G.nbunch_iter(nbunch_v))
return {
            u: sum(1 / d if d > 0 else 0 for v, d in spl(source=u).items() if v in nbunch_v)
            for u in g.nbunch_iter(nbunch)
        }
```

**Comment by dschult** [comment]
Sorry for the delay -- I've finally made time to check on this again. 

I think this proposed function need more documentation about what the method is. Harmonic centrality sums `1/dist` over all pairs of nodes `v != u`. This subset version only looks at pairs of nodes where `u` is in `nbunch` and `v` is in `nbunch_v`.  Changing the math formula might be enough, but some words would help too.

So, this measure describes something related to the centrality of `u` to a subset of nodes `nbunch_v` rather than to the whole graph. 

My next question is whether the algorithm is efficient. It computes the shortest path between each pair of nodes, starting from scratch to find the path to each `v` from `u`. Wouldn't it be better to compute shortest_paths from `u` to all the nodes `v` using a single BFS? It seems a shame to explore the early parts of a BFS over and over again. 

The full `harmonic_centrality` function computes the full BFS because it wants distances to all nodes. I would guess that is not much more costly than the somewhat shorter computation from u to v. But you only have to do it for node `u`. Whereas  this subset version has to recompute the BFS for each node `u` for each node `v`. If nbunch_v has 10 nodes, you compute 10 BFSs for each `u`. Certainly there is a tradeoff in size where it gets faster to compute all distances from `u` and only use the ones to `v`.   Does anyone know that tradeoff? 

@ChristopherReinartz how large are the subsets `nbunch_v` that you are using? and how many nodes in your graphs?

**Comment by ChristopherReinartz** [comment]
Sorry for the delay on my end, too. I can have a look at the documentation of the function this weekend. 

Regarding efficiency, I don't know the exact tradeoff. My graphs don't have too many nodes (<500) and the subsets for nodes u and v usually each contain about 30-50 nodes. 
I agree that computing all paths from the nodes u and then filtering paths to v should probably be faster.

**Comment by dschult** [comment]
For efficiency I think it is best to switch the loops so `v` is the outer loop.
```python
#  Don't reverse G anymore
centrality = {u: 0 for u in nbunch}
for v in sources:
    dist = spl(v)
    for u in nbunch:
        d = dist[u]
        if d == 0:  # handle u == v and edges with 0 weight
            continue
        centrality[u] += 1 / d
```
(untested) or similar...

It looks like that is about 10 times faster on binomial graphs with 100-300 nodes. But maybe I am leaving something out -- so check for yourself too. :}

**Comment by ChristopherReinartz** [comment]
I implemented your version of the algorithm with small modifications and it now passes all tests. It now looks like this: 

 ```python3
    nbunch = nbunch if nbunch is not None else G.nodes
    sources = sources if sources is not None else G.nodes

    spl = partial(nx.shortest_path_length, G, weight=distance)
    centrality = {u: 0 for u in nbunch}
    for v in sources:
        dist = spl(v)
        for u in set(nbunch).intersection(set(dist.keys())):
            d = dist[u]
            if d == 0:  # handle u == v and edges with 0 weight
                continue
            centrality[u] += 1 / d

    return centrality
```

File: networkx/algorithms/centrality/harmonic.py [hunk 1]
```diff
@@ -18,6 +18,11 @@ def harmonic_centrality(G, nbunch=None, distance=None):
 
     where `d(v, u)` is the shortest-path distance between `v` and `u`.
 
+    If 'sources' is given as an argument, the returned harmonic centrality
+    values are calculated as the reciprocal of the shortest path distances
+    from the nodes specified in 'sources' to 'u' instead of from all nodes
+    to 'u'.
```
Comment (dschult) [inline networkx/algorithms/centrality/harmonic.py:24]:
```suggestion
    values are calculated as the sum of the reciprocals of the shortest 
    path distances from the nodes specified in 'sources' to 'u' instead
    of from all nodes to 'u'.
```

File: networkx/algorithms/centrality/harmonic.py [hunk 2]
```diff
@@ -26,8 +31,11 @@ def harmonic_centrality(G, nbunch=None, distance=None):
       A NetworkX graph
 
     nbunch : container
```
Comment (dschult) [inline networkx/algorithms/centrality/harmonic.py:33]:
```suggestion
    nbunch : container (default: all nodes in G)
```

File: networkx/algorithms/centrality/harmonic.py [hunk 3]
```diff
@@ -26,8 +31,11 @@ def harmonic_centrality(G, nbunch=None, distance=None):
...
-      Container of nodes. If provided harmonic centrality will be computed
-      only over the nodes in nbunch.
+      Container of nodes, for which harmonic centrality values are calculated.
+
+    sources : container
+      Container of nodes. If provided, harmonic centrality will be computed
+      only with respect to paths starting at nodes in 'sources'.
```
Comment (dschult) [inline networkx/algorithms/centrality/harmonic.py:38]:
```suggestion
    sources : container (default: all nodes in G)
      Container of nodes `v` over which reciprocal distances are computed.
      Nodes not in `G` are silently ignored.
```

File: networkx/algorithms/centrality/harmonic.py [hunk 4]
```diff
@@ -54,10 +62,18 @@ def harmonic_centrality(G, nbunch=None, distance=None):
     .. [1] Boldi, Paolo, and Sebastiano Vigna. "Axioms for centrality."
            Internet Mathematics 10.3-4 (2014): 222-262.
     """
-    if G.is_directed():
-        G = G.reverse()
+
+    nbunch = nbunch if nbunch is not None else G.nodes
+    sources = sources if sources is not None else G.nodes
```
Comment (dschult) [inline networkx/algorithms/centrality/harmonic.py:67]:
```suggestion
    nbunch = set(G.nbunch_iter(nbunch)) if nbunch is not None else G.nodes
    sources = set(G.nbunch_iter(sources)) if sources is not None else G.nodes
```

File: networkx/algorithms/centrality/harmonic.py [hunk 5]
```diff
@@ -54,10 +62,18 @@ def harmonic_centrality(G, nbunch=None, distance=None):
...
+
     spl = partial(nx.shortest_path_length, G, weight=distance)
-    return {
-        u: sum(1 / d if d > 0 else 0 for v, d in spl(source=u).items())
-        for u in G.nbunch_iter(nbunch)
-    }
+    centrality = {u: 0 for u in nbunch}
+    for v in sources:
+        dist = spl(v)
+        for u in set(nbunch).intersection(set(dist.keys())):
```
Comment (dschult) [inline networkx/algorithms/centrality/harmonic.py:73]:
```suggestion
        for u in nbunch.intersection(dist):
```

**Review by dschult — COMMENTED** [review]
Excellent!   Thanks very much. This is much faster by my tests.

I think we've lost the ability to handle nodes that aren't in G. And it would be faster to avoid creating a set for nbunch inside the loop so I have a couple of suggestions and minor wording change for the docs.  Then I think we're ready to merge.

**Comment by ChristopherReinartz** [comment]
Is there anything left to do before we merge?

**Review by dschult — APPROVED** [review]

**Comment by dschult** [comment]
I was waisting time thinking about `for u in G.nodes & dist.keys():` which does work  vs `nbunch = set(G.nodes)`.
It's fine as it is.  This is ready to merge as far as I am concerned.   👍

File: networkx/algorithms/centrality/harmonic.py [hunk 6]
```diff
@@ -6,7 +6,7 @@
 __all__ = ["harmonic_centrality"]
 
 
-def harmonic_centrality(G, nbunch=None, distance=None):
+def harmonic_centrality(G, nbunch=None, sources=None, distance=None):
```
Comment (rossbar) [inline networkx/algorithms/centrality/harmonic.py:9]:
```suggestion
def harmonic_centrality(G, nbunch=None, distance=None, source=None):
```
Adding the new `sources` kwarg to the end avoids breaking the current API.

File: networkx/algorithms/centrality/harmonic.py [hunk 7]
```diff
@@ -18,16 +18,24 @@ def harmonic_centrality(G, nbunch=None, distance=None):
 
     where `d(v, u)` is the shortest-path distance between `v` and `u`.
 
+    If 'sources' is given as an argument, the returned harmonic centrality
+    values are calculated as the sum of the reciprocals of the shortest
+    path distances from the nodes specified in 'sources' to 'u' instead
+    of from all nodes to 'u'.
```
Comment (rossbar) [inline networkx/algorithms/centrality/harmonic.py:24]:
```suggestion
    If `sources` is given as an argument, the returned harmonic centrality
    values are calculated as the sum of the reciprocals of the shortest
    path distances from the nodes specified in `sources` to `u` instead
    of from all nodes to `u`.
```

File: networkx/algorithms/centrality/harmonic.py [hunk 8]
```diff
@@ -18,16 +18,24 @@ def harmonic_centrality(G, nbunch=None, distance=None):
...
     Parameters
     ----------
     G : graph
       A NetworkX graph
 
-    nbunch : container
-      Container of nodes. If provided harmonic centrality will be computed
-      only over the nodes in nbunch.
+    nbunch : container (default: all nodes in G)
+      Container of nodes, for which harmonic centrality values are calculated.
```
Comment (rossbar) [inline networkx/algorithms/centrality/harmonic.py:34]:
```suggestion
      Container of nodes for which harmonic centrality values are calculated.
```

File: networkx/algorithms/centrality/tests/test_harmonic_centrality.py [hunk 9]
```diff
@@ -92,3 +94,21 @@ def test_singleton(self):
         c = harmonic_centrality(G, distance="weight")
         d = {0: 0}
         assert c == d
+
+    def test_cycle_c4_directed(self):
+        c = harmonic_centrality(self.C4_directed, nbunch=[0, 1], sources=[1, 2])
+        d = {0: 0.833, 1: 0.333}
+        for n in sorted([0, 1]):
```
Comment (rossbar) [inline networkx/algorithms/centrality/tests/test_harmonic_centrality.py:101]:
```suggestion
        for n in [0, 1]:
```
Just a nit - sorted not necessary here.

File: networkx/algorithms/centrality/tests/test_harmonic_centrality.py [hunk 10]
```diff
@@ -92,3 +94,21 @@ def test_singleton(self):
...
+            assert almost_equal(c[n], d[n], places=3)
+
+    def test_p3_harmonic_subset(self):
+        c = harmonic_centrality(self.P3, sources=[0, 1])
+        d = {0: 1, 1: 1, 2: 1.5}
+        for n in sorted(self.P3):
```
Comment (rossbar) [inline networkx/algorithms/centrality/tests/test_harmonic_centrality.py:107]: I don't think the `sorted` would be necessary here either, unless I'm misunderstanding something!

**Review by rossbar — COMMENTED** [review]
I also apologize for taking so long to get another look at this - I think the current state is a major improvement over the previous iterations. I now understand (both from the code and docs) what the intent of the new kwarg is :)

Only one comment about the ordering of the kwargs before this is ready to go in. Other than that, just some nits/touchups that are less important. Thanks @ChristopherReinartz

**Review by rossbar — APPROVED** [review]
Thanks for the final touchups @ChristopherReinartz , this LGTM!

**Pull request merged**